### ISSUE TITLE
Develop dependency tracker for security functionality containers to ensure successful launch. 

### ISSUE THREAD

**Issue by mkbhanda** [edgexfoundry/edgex-go#1710]
Need dependency tracker to determine each container service involved in security functionality set up is launched and available.
Docker-compose's hook of "launched" is inadequate. Use the Ubuntu snaps scripts as a starting point.

Tasks to complete this:

- [x] ~~In go-mod-registry, add function to allow registering a script as a health check~~ - this is unnecessary anymore, see followup comment
- [x] in docker-edgex-consul, write script to register things that aren't go services like Kong to be registered as services in consul - this is supported via config files with consul
- [x] in edgex-go modify the security-secrets-setup container to wait for security-secrets-setup to be done before running vault (note we will defer splitting security-secrets-setup into its own container separate from vault to later) (also note this is basically already done)
- [x] in docker-edgex-consul, write consul health script for kong-migration - https://github.com/edgexfoundry/docker-edgex-consul/pull/2
- [x] in docker-edgex-consul, write consul health script for kong-database - https://github.com/edgexfoundry/docker-edgex-consul/pull/2
- [x] in docker-edgex-consul, write consul health script for kong - https://github.com/edgexfoundry/docker-edgex-consul/pull/2
- [x] in docker-edgex-volume, register services kong, kong-migration, kong-database as services in consul with local health scripts defined - this is supported via config files with consul - https://github.com/edgexfoundry/docker-edgex-volume/pull/9
- [x] in docker-edgex-volume, register services such as security-secrets-setup, security-proxy-setup, and security-secretstore-setup with local health scripts defined - this is supported via config files with consul
- [x] in docker-edgex-consul, write consul health script for security-secrets-setup and define in consul config file
- [x] in docker-edgex-consul, write consul health script for security-secretstore-setup and define in consul config file
- [x] in docker-edgex-consul, write consul health script for security-proxy-setup and define in consul config file
- [x] In go-mod-secrets, update to wait for the service known as `security-secretstore-setup` in consul is healthy before accessing vault
- [x] update edgex-go to use new version of go-mod-secrets

**Comment by mkbhanda** [comment]
![edgex_security_dependancies](https://user-images.githubusercontent.com/2702293/63894441-7e749d00-c9a1-11e9-92d8-f3914ee581d1.png)

**Comment by bnevis-i** [comment]
Added backreferences to #1676 and #1717 -- the Vault access tokens must be generated and initial secrets must be loaded before we can signal to other services that Vault is online.

**Comment by mkbhanda** [comment]
The above figure from Security F2F, Aug 1, 2019 Intel/Santa Clara. Capturing my understanding below:
1) Consul must first come up. It will hold the state of other services as they come up.
2) Vault (current secret store implementation) and Kong (current security proxy) can both be launched in parallel.

Both Vault and Kong need PKI, public-private key pair and certificates to be able to support secure communications. Vault so that secrets are not leaked in transit. Kong so that API calls are encrypted and protected.  The PKI materials could be created on the fly or imported. 

3) Once  Consul indicates Vault has successfully launched Initialize PKI -- creating/loading PKI materials in Vault. Configure Vault server certificate using above materials. Based on PKI initialization arguments possibly store or destroy CA private key.

4) Once Consult indicates Vault PKI initialization completed, Kong may communicate with Vault server to retrieve its server certificate.

5) Consul status "Kong initialized" completes the minimal security steps.

6) All other services as part of their secrets (if any) download would communicate with Vault once Consul indicates Vault initialized. The security clients would need the CA certificate, and their access token. Bryon investigating whether token and certs necessary .. (Slack dialog between Ian Johnson and Bryon Nevis)

4) 
3b) PKI setup Vault needs to check that Vault is "up" in Consul

**Comment by anonymouse64** [comment]
Note that the complexity of this task grows if we take @bnevis-i's suggestion to split up `security-secretstore-setup` into separate processes

**Comment by anonymouse64** [comment]
As per the WG meeting last week, I will be taking this on, and in addition we will not be splitting up `security-secretstore-setup` into multiple binaries for each individual task, instead we will be implementing a configuration option which allows specifying executables to run for some of the different tasks. A full breakup/customization will be discussed for Geneva at some point in the future.

**Comment by anonymouse64** [comment]
Continuing on from @mkbhanda's comment above, we have the following ordering dependencies (with `->` meaning "depends on"):

* Vault → Consul
* Vault → `security-secrets-setup`
* Kong → `kong-database` (Postgres or Cassandra)
* Kong → `kong-migrations`
* `security-proxy-setup` → Kong
* `security-proxy-setup` → `security-secrets-setup`
* `security-secretstore-setup` → Vault
* `security-secretstore-setup` → `security-secrets-setup`

This means we have the following set of services which will implement their Consul health checks as follows:

### `security-secrets-setup`

* Check for existence of public PKI certificate files used by Kong and Vault at a configurable location

### Vault

* Wait for `security-secrets-setup` to be healthy as reported by Consul

### Kong

* Wait for `kong-database` to be healthy as reported by Consul
* Wait for `kong-migrations` to be healthy as reported by Consul
* Wait for nginx/Kong itself to respond on it's admin API (or some other mechanism of determining that Kong is actually listening on it's ports, etc.)

### `security-proxy-setup`

* Wait for Kong to be healthy as reported by Consul
* Wait for `security-secrets-setup` to be healthy as reported by Consul

### `security-secretstore-setup`

* Wait for Vault `/sys/health` endpoint to report 200 (meaning that Vault is fully active, initialized and unsealed
* Wait for `security-secrets-setup` to be healthy as reported by Consul

Note that we don't consider Vault to be "ready to use" for the rest of EdgeX until `security-secretstore-setup` has finished, so things like go-mod-secrets will need to wait on `security-secretstore-setup` and not on Vault proper. Also note that AFAICT, there's no need for go-mod-secrets to wait on `security-proxy-setup` since currently all intra-EdgeX communication is done without the proxy.

**Comment by mkbhanda** [comment]
@Tingyu and I were talking about creating files as each dependency milestone was achieved if we want to take the sentinel file(s) approach .. on a shared persistent volume.
or example "value-unsealed.txt", "pki_setup_complete.txt", kong_certificate_loaded.txt, vault_certificate_loaded.txt"

**Comment by anonymouse64** [comment]
@mkbhanda I think we should avoid using files as much as possible because they are very prone to race conditions and making it not subject to race conditions requires things like [`flock`](https://linux.die.net/man/2/flock) which are difficult to port (but I'm happy to hear a simple solution to this).

However additionally, making things dependent on files is problematic when we start looking at having multiple instances and having services spread out across machines in k8s like deployments.

**Comment by mkbhanda** [comment]
@anonymouse64 totally aware of the dangers of file based solutions with respect to locking and in distributed environments. For the distributed/K8 environment was thinking of files on a persistent volume that the various containers involved would need to mount. To avoid synchronization issues pertaining to write to a single status file and parsing the same to determine if an item of interest exists to instead use a file per status we are tracking. Essentially the volume behaves like a blackboard and each status we are interested is a mark on the blackboard. First step of EdgeX launch is to clear this persistent volume. 

But to maintain consistency with the container level status checking supported by Consul, if we could instead write directly to consul after each of these sub-steps, that would be better. A state "push" to Consul versus a call-back pull function from Consul to update state.
If Consul only provides pull type status update, then the files on a persistent volume might be necessary.

**Comment by anonymouse64** [comment]
It doesn't appear as if Consul has an explicit API for pushing the health status of a service, however it does allow you to specify the initial health of a service when setting up the health check, so perhaps it is possible to just define the same health check for a service but with the status set to healthy when that happens. 

The health check interval can be configured, so perhaps we can configure the time of how often the health check happens for these services specifically to happen very frequently while things are being setup, then happen more slowly when they are all setup.

**Comment by anonymouse64** [comment]
Things to do to finish this:

- [] Add function to go-mod-registry to allow registering a script as a health check 
- [] Update go-mod-secrets to wait for the service known as `security-secretstore-setup` in consul is healthy before accessing vault
- [] write script to register things that aren't go services like Kong to be registered as services in consul
- [] write consul health script for kong-migration
- [] write consul health script for kong-database
- [] write consul health script for kong
- [] write consul health script for security-secrets-setup and reference this in the call to go-mod-registry when it starts up
- [] write consul health script for security-secretstore-setup and reference this in the call to go-mod-registry when it starts up
- [] write script to wait for security-secrets-setup to be done before running vault (note it's a bit unclear to me how we can do this while still using the upstream vault container w/o modifications)
- [] write consul health script for security-proxy-setup

**Comment by anonymouse64** [comment]
We no longer need to modify go-mod-registry or edgex-go, since Consul supports defining checks (and services) through config files, so what we can do for Fuji with the least amount of work is just define the services for security components and health scripts via configuration files inside the docker container.

This means consul will always startup knowing about the security services and how to check if they are healthy.

It also means that for the docker-compose setup without security we will just use the upstream consul docker image, and for security use the docker-edgex-consul version with the health check scripts and config files setting those up.

The only remaining Go code task here will be setting up go-mod-secrets to block on `security-secretstore-setup` to be healthy before continuing. 

The other reason we should implement this through static config files defining the script instead of the HTTP API is that the HTTP admin API consul exposes effectively allows remote code execution attacks if Consul is exposed on the network. See https://www.hashicorp.com/blog/protecting-consul-from-rce-risk-in-specific-configurations for more details.

**Comment by anonymouse64** [comment]
Some PR's opened to start addressing this:
- https://github.com/edgexfoundry/docker-edgex-consul/pull/2
- https://github.com/edgexfoundry/docker-edgex-volume/pull/9

Will also open up a developer-scripts PR using these containers as well

**Comment by anonymouse64** [comment]
developer-scripts PR is here: https://github.com/edgexfoundry/developer-scripts/pull/147

See the info on that PR for how to test the other PRs

**Comment by jpwhitemn** [comment]
@mkbhanda @anonymouse64 @bnevis-i - should what remains of this issue be moved to Geneva tag or is this still all Fuji work?

**Comment by anonymouse64** [comment]
Unclear if we can still do this for Fuji, I still haven't gotten a review on https://github.com/edgexfoundry/developer-scripts/pull/147

**Comment by michaelestrin** [comment]
https://github.com/edgexfoundry/developer-scripts/pull/147 has been reviewed and only needs to be merged.  Merging now.

**Comment by anonymouse64** [comment]
Cool beans! :tada: 

I will open up the followup PR's this weekend most likely, I have other things to get to today unfortunately

**Comment by anonymouse64** [comment]
The last PR's for this are open:
- https://github.com/edgexfoundry/docker-edgex-volume/pull/11
- https://github.com/edgexfoundry/docker-edgex-consul/pull/6
- https://github.com/edgexfoundry/edgex-go/pull/2075
- https://github.com/edgexfoundry/developer-scripts/pull/172

I have omitted the go-mod-secrets change, let's discuss this in person at the F2F and I will comment here with what we decide there.

**Comment by anonymouse64** [comment]
See also https://github.com/edgexfoundry/developer-scripts/issues/167#issuecomment-549951096

**Comment by anonymouse64** [comment]
Final PR's here: 
- https://github.com/edgexfoundry/docker-edgex-mongo/pull/59
- https://github.com/edgexfoundry/go-mod-secrets/pull/36

Home stretch folks for Fuji! :taco:

**Comment by anonymouse64** [comment]
Well 2 more PR's here to get edgex-mongo to behave and not be racy:

- https://github.com/edgexfoundry/developer-scripts/pull/175
- https://github.com/edgexfoundry/docker-edgex-consul/pull/7
- https://github.com/edgexfoundry/docker-edgex-mongo/pull/59

**Comment by anonymouse64** [comment]
Last PR that needs to be merged before we can close this is https://github.com/edgexfoundry/developer-scripts/pull/175

**Comment by anonymouse64** [comment]
Forward-port PR's to Geneva/master are up:
- https://github.com/edgexfoundry/docker-edgex-volume/pull/12
- https://github.com/edgexfoundry/docker-edgex-consul/pull/9
- https://github.com/edgexfoundry/docker-edgex-consul/pull/8

**Comment by anonymouse64** [comment]
Moving this task to Geneva, since all the Fuji PR's/tasks have been completed

**Comment by anonymouse64** [comment]
Last PR's for this for Geneva that are still open:
- https://github.com/edgexfoundry/docker-edgex-volume/pull/12
- https://github.com/edgexfoundry/docker-edgex-consul/pull/8

**Comment by bnevis-i** [comment]
Final validation of the geneva nightlys (and closing out this issue) is blocked by https://github.com/edgexfoundry/docker-edgex-consul/issues/13.

**Comment by bnevis-i** [comment]
https://github.com/edgexfoundry/docker-edgex-consul/issues/13 is now closed.  Closing this issue.
